File: app/src/main/java/workshop/akbolatss/xchangesrates/repositories/ChartRepository.java
package workshop.akbolatss.xchangesrates.repositories;

import java.util.List;

import io.reactivex.Single;
import workshop.akbolatss.xchangesrates.model.dao.Snapshot;
import workshop.akbolatss.xchangesrates.model.dao.SnapshotChart;
import workshop.akbolatss.xchangesrates.model.dao.SnapshotInfo;
import workshop.akbolatss.xchangesrates.model.response.ChartResponse;

/**
 * Author: Akbolat Sadvakassov
 * Date: 04.01.2018
 */

public interface ChartRepository {

    Single<ChartResponse> getQueryResult(String coin, String exchange,
                                         String currency, String term);

    Single<List<Snapshot>> getAllChartData();

    Single<List<Snapshot>> getActiveChartData();

    Single<SnapshotInfo> getChartDataInfo(long key);

    Single<ChartResponse> getSnapshot(String coin, String exchange,
                                      String currency, String term);

    Single<Boolean> onAddChartData(Snapshot data, SnapshotInfo dataInfo, List<SnapshotChart> dataCharts);

    Single<Integer> onDeleteChartData(long key);

    void onUpdateChartData(Snapshot snapshot, SnapshotInfo dataInfo, List<SnapshotChart> chartsList);

    Single<Integer> onOptionsChanged(long key, boolean isActive, String timing);
}

File: app/src/main/java/workshop/akbolatss/xchangesrates/repositories/ChartRepository.kt
package workshop.akbolatss.xchangesrates.repositories

import io.reactivex.Single
import workshop.akbolatss.xchangesrates.model.dao.Snapshot
import workshop.akbolatss.xchangesrates.model.dao.SnapshotChart
import workshop.akbolatss.xchangesrates.model.dao.SnapshotInfo
import workshop.akbolatss.xchangesrates.model.response.ChartResponse

/**
 * Author: Akbolat Sadvakassov
 * Date: 04.01.2018
 */

interface ChartRepository {

    val allChartData: Single<List<Snapshot>>

    val activeChartData: Single<List<Snapshot>>

    fun getQueryResult(coin: String, exchange: String,
                       currency: String, term: String): Single<ChartResponse>

    fun getChartDataInfo(key: Long): Single<SnapshotInfo>

    fun getSnapshot(coin: String, exchange: String,
                    currency: String, term: String): Single<ChartResponse>

    fun onAddChartData(data: Snapshot, dataInfo: SnapshotInfo, dataCharts: List<SnapshotChart>): Single<Boolean>

    fun onDeleteChartData(key: Long): Single<Int>

    fun onUpdateChartData(snapshot: Snapshot, dataInfo: SnapshotInfo, chartsList: List<SnapshotChart>)

    fun onOptionsChanged(key: Long, isActive: Boolean, timing: String): Single<Int>
}
